File: typescript/src/workshop/m8/s12_expandcontract/step1/BookingRepository.ts
import type { Booking } from '../Booking.js';
import { type BookingTable, Row } from '../BookingTable.js';
import { BookingPayloadFormat } from './BookingPayloadFormat.js';
import { CsvBookingFormat } from './CsvBookingFormat.js';

/**
 * Krok 1: expand + dual write - zapis do obu kolumn, odczyt nadal ze starej. Wycofanie do
 * poprzedniej wersji jest bezpieczne: stara wersja czyta csv, który wciąż powstaje.
 */
export class BookingRepository {
  constructor(private readonly table: BookingTable) {}

  save(booking: Booking): void {
    this.table.put(booking.id, new Row(CsvBookingFormat.write(booking),
      BookingPayloadFormat.write(booking)));
  }

  find(id: string): Booking | undefined {
    const csv = this.table.get(id)?.csv;
    return csv == null ? undefined : CsvBookingFormat.read(csv);
  }
}

File: csharp/src/Training.Workshop/M8/S12ExpandContract/Step1/BookingRepository.cs
namespace Training.Workshop.M8.S12ExpandContract.Step1;

/// <summary>
/// Krok 1: expand + dual write - zapis do obu kolumn, odczyt nadal ze starej. Wycofanie do
/// poprzedniej wersji jest bezpieczne: stara wersja czyta Csv, który wciąż powstaje.
/// </summary>
public sealed class BookingRepository
{
    private readonly BookingTable _table;

    public BookingRepository(BookingTable table)
    {
        _table = table;
    }

    public void Save(Booking booking)
    {
        _table.Put(booking.Id, new BookingTable.Row(CsvBookingFormat.Write(booking),
            BookingPayloadFormat.Write(booking)));
    }

    public Booking? Find(string id)
    {
        string? csv = _table.Get(id)?.Csv;
        return csv is null ? null : CsvBookingFormat.Read(csv);
    }
}
